_posts/2019-01-15-recursion.md

```diff
@@ -0,0 +1,313 @@
+---
+author: Sophie DeBenedetto
+author_link: https://github.com/sophiedebenedetto
+categories: recursion
+date: 2019-01-15
+layout: post
+title: Understanding Recursion with Elixir
+excerpt: >
+  De-mystify the concept of recursion and gain a deeper understanding of how and why to use it by writing our very own recursive function in Elixir.
+---
+
+"Recursion" can be a scary word for those of us unfamiliar with its applications. In this post, we'll de-mystify the concept of recursion and gain a deeper understanding of how and why to use it by writing our very own recursive function in Elixir.
+
+## What is Recursion
+
+In short, "recursion" is when a function calls itself. First we'll look at a contrived example. Later on in this post we'll build a more practical recursive function.
+
+Below we've defined a function `RecursionPractice.hello_world/0`, that calls itself:
+
+```elixir
+defmodule RecursionPractice do
+  def hello_world do
+    IO.puts "Hello, World!"
+    hello_world()
+  end
+end
+```
+
+If you think that invoking our `RecursionPractice.hello_world/0` function would cause `"Hello, World!"` to get puts'd out to the terminal infinitely--you're right! The `hello_world` function does two things:
+
+1. Puts out "Hello, World!"
+2. Call `hello_world` (again)
+
+When `hello_world` is invoked again, it will do two things:
+
+1. Puts out "Hello, World!"
+2. Call `hello_world` (again)
+
+And so on. While "a function that calls itself" is the basic definition of recursion, its _not_ how we want to implement a recursive function.
+
+Any recursive function needs a way to _stop calling itself_ under a certain condition. This condition is often referred to as the **base case**. Let's create a base case for our `hello_world` function. We'll count the number of times we call the function and stop calling it once we've reached 10.
+
+```elixir
+def hello_world(count \\ 0) do
+  IO.puts "Hello, World!"
+  if count < 10 do
+    new_count = count + 1
+    hello_world(new_count)
+  end
+end
+```
+
+The `if` condition controls our recursive function for us. _If_ the count is less than 10, increment the count by 1 and call `hello_world` again. Otherwise, _don't do anything_, i.e. stop calling the recursive function!
```

Comment:
Make sure to use function/arity format when referencing functions: `hello_world/0`